validation/src/main/java/io/micronaut/validation/routes/RouteValidationVisitor.java

```diff
@@ -111,10 +112,44 @@ public void visitMethod(MethodElement element, VisitorContext context) {
 
     @Override
     public void start(VisitorContext visitorContext) {
-        String prop = visitorContext.getOptions().getOrDefault(VALIDATION_OPTION, "true");
-        skipValidation = prop != null && prop.equals("false");
+        skipValidation = shouldSkipRouteValidation(visitorContext);
         rules.add(new MissingParameterRule());
         rules.add(new NullableParameterRule());
         rules.add(new RequestBeanParameterRule());
     }
+
+    /**
+     * Check whether to skip route validation.
+     * <p>
+     * Route validation is disabled when using Java 8 or below with incremental compilation. The
+     * Java 8 compiler does not load parameter names for compiled classes, so in this case it is
+     * impossible to match route parameters to method parameters.
+     *
+     * @param visitorContext The visitor context
+     * @return A boolean indicating whether to skip route validation.
+     */
+    private static boolean shouldSkipRouteValidation(VisitorContext visitorContext) {
+        int javaVersion = getVersion();
+        if (javaVersion < 9) {
+            String incremental = visitorContext.getOptions().get(MICRONAUT_PROCESSING_INCREMENTAL);
+            if (incremental != null && incremental.equals("true")) {
+                return true;
+            }
+        }
+        String prop = visitorContext.getOptions().getOrDefault(VALIDATION_OPTION, "true");
+        return prop != null && prop.equals("false");
+    }
+
+    private static int getVersion() {
+        String version = System.getProperty("java.version");
+        if (version.startsWith("1.")) {
+            version = version.substring(2, 3);
+        } else {
+            int dot = version.indexOf(".");
+            if (dot != -1) {
+                version = version.substring(0, dot);
+            }
+        }
+        return Integer.parseInt(version);
```

Comment:
This could lead to a NPE if one of the conditions fail. They shouldn't fail, but you never know. Best to wrap in an a try/catch and return -1 if an exception occurs.